src/utils/convertTimeStringToObj.ts

```diff
@@ -1,36 +1,50 @@
 import { TimeObjectSchema } from 'serina.schema';
-import { MERIDIEM, PM } from 'filters/time/time.constants';
+import TIME from 'filters/time/time.constants';
 import contains from './contains';
-import { remove, isValidTime } from 'utils';
+import { matchPattern } from 'utils';
 
-function convertTimeStringToObj(time: string): TimeObjectSchema {
+function getValidMatch(text: string, pattern: string): string {
+    const matched = matchPattern(text, pattern, false);
+    if (!matched || matched.length === 0) return null;
+    const [ value ] = matched;
+    return value;
+}
+
+function convertTimeStringToObj(timeString: string): TimeObjectSchema {
     let hour;
     let minute;
-
-    if (contains(time, `${MERIDIEM}`, false)) {
-        // 12H
-        const timeWithMeridiem = remove(time, `${MERIDIEM}`, false);
-        const [ hourStr, minStr ] = timeWithMeridiem.split(':');
-        hour = parseInt(hourStr, 10);
-        minute = parseInt(minStr, 10);
-
-        if (isNaN(minute)) minute = 0;
-
-        if (contains(time, `${PM}`,false)) {
-            // Post Meridiem (PM)
-            if (hour < 12) hour += 12;
-        } else {
-            // Ante Meridiem (AM)
-            if (hour === 12) hour = 0;
+    
+    const isRelativeTime = contains(timeString, TIME.RELATIVE_TIME_FILLER_WORDS);
+    
+    if (isRelativeTime) {
+        if (contains(timeString, TIME.VERBAL_QUANTIFIERS)) { // half or quarter
+            hour = getValidMatch(timeString, `(?<=${TIME.RELATIVE_TIME_FILLER_WORDS}( ))${TIME.HOUR_PART}`)
+            if (contains(timeString, TIME.HALF)) minute = '30';
+            if (contains(timeString, TIME.QUARTER)) minute = '15';
+        } else { // 20 min to 7pm
+            hour = getValidMatch(timeString, `(?<=${TIME.RELATIVE_TIME_FILLER_WORDS}( ))${TIME.HOUR_PART}(?=((( )${TIME.MERIDIEM}))?)`);
+            minute = getValidMatch(timeString, `\\b${TIME.MINUTE_PART}(?=(( )(${TIME.MINUTE_IDENTIFIER}( ))?${TIME.RELATIVE_TIME_FILLER_WORDS}))`);
```

Comment:
I know this is shorter than splitting the string up, but is it really more readable? Very hard without a debugger to figure out how this works, since it creates a pretty complicated regex.

Reply:
I am not sure either, I think the old approach was too hard to visualise what is going on. With this at least you see in one line what is happening. Again we can revisit this when we refactor further down the line, I'm not 100% happy with this either. Plus the `?<=` support issue means it is it is going to be problematic as well.